From 40a64399edf3a48b6dab111118f89755afdf4664 Mon Sep 17 00:00:00 2001
From: Ruoyu Ying <ruoyu.ying@intel.com>
Date: Fri, 5 May 2023 03:04:48 -0400
Subject: [PATCH 05/16] ima: use SHA384 hash algorithm for measurement in
 kernel sfs

Kernel IMA by default uses SHA1 hash algorithm to show the measurements
stored inside kernel security fs. Due to the fact the TDX RTMR only
supports SHA384, this patch will change the output of the measurements
within kernel security fs into SHA384 hash.
---
 security/integrity/ima/ima_fs.c | 15 +++++++++++----
 1 file changed, 11 insertions(+), 4 deletions(-)

diff --git a/security/integrity/ima/ima_fs.c b/security/integrity/ima/ima_fs.c
index e2aee9553246..85f2f3a75191 100644
--- a/security/integrity/ima/ima_fs.c
+++ b/security/integrity/ima/ima_fs.c
@@ -150,8 +150,11 @@ int ima_measurements_show(struct seq_file *m, void *v)
 	pcr = !ima_canonical_fmt ? e->pcr : (__force u32)cpu_to_le32(e->pcr);
 	ima_putc(m, &pcr, sizeof(e->pcr));
 
-	/* 2nd: template digest */
-	ima_putc(m, e->digests[ima_sha1_idx].digest, TPM_DIGEST_SIZE);
+	/* 2nd: template digest, use SHA384 instead of SHA1 in TDX RTMR case */
+	if (ima_tdx_device)
+		ima_putc(m, e->digests[ima_hash_algo_idx].digest, SHA384_DIGEST_SIZE);
+	else
+		ima_putc(m, e->digests[ima_sha1_idx].digest, TPM_DIGEST_SIZE);
 
 	/* 3rd: template name size */
 	namelen = !ima_canonical_fmt ? strlen(template_name) :
@@ -233,8 +236,12 @@ static int ima_ascii_measurements_show(struct seq_file *m, void *v)
 	/* 1st: PCR used (config option) */
 	seq_printf(m, "%2d ", e->pcr);
 
-	/* 2nd: SHA1 template hash */
-	ima_print_digest(m, e->digests[ima_sha1_idx].digest, TPM_DIGEST_SIZE);
+	/* 2nd: SHA1 template hash, use SHA384 instead of SHA1 in TDX RTMR case */
+	if (ima_tdx_device) {
+		ima_print_digest(m, e->digests[ima_hash_algo_idx].digest, SHA384_DIGEST_SIZE);
+	} else {
+		ima_print_digest(m, e->digests[ima_sha1_idx].digest, TPM_DIGEST_SIZE);
+	}
 
 	/* 3th:  template name */
 	seq_printf(m, " %s", template_name);
-- 
2.40.1

